design/baremetal-operator/bmh_non-bootable_iso.md

```diff
@@ -0,0 +1,153 @@
+to attach a non-bootable iso image to hosts.
+
+We may consider adding support for HTTP credentials and TLS settings in the future
+but it's outside the scope of this proposal.
+
+**Note** : As part of this proposal, the dataImage will be attached after the node
+has been Provisioned or ExternallyProvisioned.
+
+### Ironic Support
+
+The [proposal](https://bugs.launchpad.net/ironic/+bug/2033288) for exposing this functionaility in Ironic has been
+accepted by the upstream Ironic community.
+
+The [Ironic implementation](https://review.opendev.org/c/openstack/ironic/+/894918) based on the above proposal has been completed.
+
+Other than that, we just need the implementation for driver specific support,
+which in this proposal is Redfish.
+
+## Design Details
+
+Create a new Custom Resource(CR), `DataImage`, with spec field `Url` and the same `Name`
+and `Namespace` as the `BareMetalHost` where we are attaching the image.
+
+To attach an ISO, first the `DataImage` CR object is created and then it is attached to the
+BMH on its next reboot/power on.
+
+The `DataImage` `Status` will cache the image details including the attachment
+status, which will inform us if the attachment succeeded and can also be used
+for detachment - either when the attachment fails and we retry or when the user
+explicitly requests detachment by deleting the `DataImage` CRD that corresponds to a BMH.
+
+We will introduce a new flag, `NonBootableISOImage` (under `pkg/hardwareutils/bmc`), at the
+webhook level to validate if a given driver supports the non-bootable ISO feature.
+
+### Implementation Details
+
+As part of the design we are assuming that to attach an image to a BMH object, the user
+will create the corresponding `DataImage` CR (with the same `Name` and `Namespace` as the
+`BareMetalHost` object). The BMH Controller will add the ownerReference for the corresponding BMH
+(example : [HostFirmwareSettings](https://github.com/metal3-io/baremetal-operator/blob/9ce684e6462a6ab55ff650cb6c11f9ba1ffb395d/docs/api.md?plain=1#L664)) and it will also add a finalizer to the `DataImage` after its attached.
+Then, on the next reboot/power-on of the BMH, the `DataImage` will be attached and the `DataImage`
+`Status` will be updated to reflect the status of the attachment, including which image was attached.
+In case of detachment of the ISO(`DataImage`), similar process will be followed where the actual
+detachment will happen on the next reboot/power-on and the `DataImage` Status will reflect the status
+of detachment, example : number of failures. The finalizer will also be removed once the
+detachment succeeds.
+Here is an example of what the `DataImage` `Status` might look like:
+
+```yaml
+status:
+  lastReconciliation: "2024-01-01T12:00:00Z"
+  error:
+    count: 0
+    message: ""
+  attachedImage:
+    url: "http://example.com/images/dataimage.iso"
+```
+
+Since the attachment of the ISO happens when a BMH reconcile is triggered as a result of BMH
```

Comment:
Is there btw some other way to force  attach the ISO without booting/rebooting the machine? Or should we have a way if we don't have one to make the BMH reconcile and attach the medua without reboot? Wouldn't it be better if we would have an annotation, something like "attachIso" to start reconciliation instead of requiring a reboot? It feels a bit excessive to require a full reboot when a non bootable ISO image is attached.

Reply:
It's somewhat risky. A lot of hardware will not attach a device without a reboot. iDRAC tends to operate in terms of tasks: once you create a task to attach a virtual media, this task will be pending until the next reboot. It may interfere with later tasks, e.g. if you try to clean the host.

Maybe we can later extend the DataImage object with something like `immediate: true`? In any case, I'd not include it into the first version, if that's acceptable for you.

Reply:
Thank you for the explanation @dtantsur , I am fine with this reasoning but I would like to ask you @hroyrh  to please include this reasoning in the proposal for future reference and please mention that in the future we could implement a method that would allow us to attach block device without restarting, please also include the iDRAC example also.